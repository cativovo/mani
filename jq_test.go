package main

import (
	"encoding/json"
	"fmt"
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

func TestRunQuery(t *testing.T) {
	bulbasaur := getJSON(t, "bulbasaur.json")
	charmander := getJSON(t, "charmander.json")
	squirtle := getJSON(t, "squirtle.json")
	gen1Starters := getJSON(t, "gen1-starters.json")

	testCases := []struct {
		name         string
		json         string
		filter       string
		flags        JQFlags
		validateJSON bool
	}{
		{
			name:         "valid filter",
			json:         bulbasaur,
			filter:       ".name",
			validateJSON: true,
		},
		{
			name:         "invalid filter",
			json:         bulbasaur,
			filter:       "jkjkjkjkjkjk",
			validateJSON: false,
		},
		{
			name:         "indent array input",
			json:         gen1Starters,
			filter:       ".",
			validateJSON: true,
		},
		{
			name:         "indent key value input",
			json:         bulbasaur,
			filter:       ".",
			validateJSON: true,
		},
		{
			name:         "unpack array",
			json:         gen1Starters,
			filter:       ".[]",
			validateJSON: false,
		},
		{
			name:   "compact array input",
			json:   gen1Starters,
			filter: ".",
			flags: JQFlags{
				Compact: true,
			},
			validateJSON: true,
		},
		{
			name:   "compact key value input",
			json:   bulbasaur,
			filter: ".",
			flags: JQFlags{
				Compact: true,
			},
			validateJSON: true,
		},
		{
			name:   "compact unpack array",
			json:   gen1Starters,
			filter: ".[]",
			flags: JQFlags{
				Compact: true,
			},
			validateJSON: false,
		},
		{
			name:   "cooked",
			json:   gen1Starters,
			filter: ".[].name",
			flags: JQFlags{
				Raw: false,
			},
			validateJSON: false,
		},
		{
			name:   "it's fucking raw",
			json:   gen1Starters,
			filter: ".[].name",
			flags: JQFlags{
				Raw: true,
			},
			validateJSON: false,
		},
		{
			name:   "slurp",
			json:   bulbasaur + charmander + squirtle,
			filter: ".",
			flags: JQFlags{
				Slurp: true,
			},
			validateJSON: true,
		},
		{
			name:   "compact and slurp",
			json:   bulbasaur + charmander + squirtle,
			filter: ".",
			flags: JQFlags{
				Compact: true,
				Slurp:   true,
			},
			validateJSON: true,
		},
		{
			name:   "raw and slurp",
			json:   bulbasaur + charmander + squirtle,
			filter: ".[].name",
			flags: JQFlags{
				Raw:   true,
				Slurp: true,
			},
			validateJSON: false,
		},
		{
			name:   "invalid json",
			json:   `{"yot":}`,
			filter: ".",
		},
		{
			name: "invalid json with slurp",
			json: `{"yot":true}{"foo":}`,
			flags: JQFlags{
				Slurp: true,
			},
			filter: ".",
		},
	}

	for _, testCase := range testCases {
		t.Run(testCase.name, func(t *testing.T) {
			result := RunQuery(testCase.json, testCase.filter, testCase.flags)

			if testCase.validateJSON && !json.Valid([]byte(result)) {
				t.Error("produced an invalid json")
			}

			snaps.MatchSnapshot(t, result)
		})
	}
}

func getJSON(t *testing.T, filename string) string {
	t.Helper()
	data, err := os.ReadFile(fmt.Sprintf("testdata/%s", filename))
	if err != nil {
		t.Fatal(err)
	}

	return string(data)
}
