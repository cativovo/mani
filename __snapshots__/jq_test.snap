
[TestRunQuery/valid_filter - 1]
"Bulbasaur"
---

[TestRunQuery/invalid_filter - 1]
function not defined: jkjkjkjkjkjk/0
---

[TestRunQuery/array_input - 1]
[{"name":"bulbasaur","url":"https://pokeapi.co/api/v2/pokemon/1/"},{"name":"ivysaur","url":"https://pokeapi.co/api/v2/pokemon/2/"},{"name":"venusaur","url":"https://pokeapi.co/api/v2/pokemon/3/"},{"name":"charmander","url":"https://pokeapi.co/api/v2/pokemon/4/"},{"name":"charmeleon","url":"https://pokeapi.co/api/v2/pokemon/5/"},{"name":"charizard","url":"https://pokeapi.co/api/v2/pokemon/6/"},{"name":"squirtle","url":"https://pokeapi.co/api/v2/pokemon/7/"},{"name":"wartortle","url":"https://pokeapi.co/api/v2/pokemon/8/"},{"name":"blastoise","url":"https://pokeapi.co/api/v2/pokemon/9/"}]
---

[TestRunQuery/indent_array_input - 1]
[
  {
    "name": "bulbasaur",
    "url": "https://pokeapi.co/api/v2/pokemon/1/"
  },
  {
    "name": "ivysaur",
    "url": "https://pokeapi.co/api/v2/pokemon/2/"
  },
  {
    "name": "venusaur",
    "url": "https://pokeapi.co/api/v2/pokemon/3/"
  },
  {
    "name": "charmander",
    "url": "https://pokeapi.co/api/v2/pokemon/4/"
  },
  {
    "name": "charmeleon",
    "url": "https://pokeapi.co/api/v2/pokemon/5/"
  },
  {
    "name": "charizard",
    "url": "https://pokeapi.co/api/v2/pokemon/6/"
  },
  {
    "name": "squirtle",
    "url": "https://pokeapi.co/api/v2/pokemon/7/"
  },
  {
    "name": "wartortle",
    "url": "https://pokeapi.co/api/v2/pokemon/8/"
  },
  {
    "name": "blastoise",
    "url": "https://pokeapi.co/api/v2/pokemon/9/"
  }
]
---

[TestRunQuery/indent_key_value_input - 1]
{
  "abilities": [
    {
      "description": "Powers up Grass-type moves when the Pokémon's HP is low.",
      "name": "Overgrow"
    },
    {
      "description": "Boosts the Pokémon's Speed stat in sunny weather.",
      "hidden": true,
      "name": "Chlorophyll"
    }
  ],
  "description": "Bulbasaur can be seen napping in bright sunlight. There is a seed on its back. By soaking up the sun's rays, the seed grows progressively larger.",
  "evolution": {
    "evolves_to": {
      "id": 2,
      "name": "Ivysaur"
    },
    "level": 16
  },
  "id": 1,
  "moves": [
    {
      "accuracy": 100,
      "name": "Tackle",
      "power": 40,
      "pp": 35,
      "type": "Normal"
    },
    {
      "accuracy": 100,
      "name": "Vine Whip",
      "power": 45,
      "pp": 25,
      "type": "Grass"
    },
    {
      "accuracy": 90,
      "name": "Leech Seed",
      "power": null,
      "pp": 10,
      "type": "Grass"
    },
    {
      "accuracy": 100,
      "name": "Growl",
      "power": null,
      "pp": 40,
      "type": "Normal"
    }
  ],
  "name": "Bulbasaur",
  "stats": {
    "attack": 49,
    "defense": 49,
    "hp": 45,
    "special_attack": 65,
    "special_defense": 65,
    "speed": 45
  },
  "type": [
    "Grass",
    "Poison"
  ]
}
---

[TestRunQuery/compact_array_input - 1]
[{"name":"bulbasaur","url":"https://pokeapi.co/api/v2/pokemon/1/"},{"name":"ivysaur","url":"https://pokeapi.co/api/v2/pokemon/2/"},{"name":"venusaur","url":"https://pokeapi.co/api/v2/pokemon/3/"},{"name":"charmander","url":"https://pokeapi.co/api/v2/pokemon/4/"},{"name":"charmeleon","url":"https://pokeapi.co/api/v2/pokemon/5/"},{"name":"charizard","url":"https://pokeapi.co/api/v2/pokemon/6/"},{"name":"squirtle","url":"https://pokeapi.co/api/v2/pokemon/7/"},{"name":"wartortle","url":"https://pokeapi.co/api/v2/pokemon/8/"},{"name":"blastoise","url":"https://pokeapi.co/api/v2/pokemon/9/"}]
---

[TestRunQuery/compact_key_value_input - 1]
{"abilities":[{"description":"Powers up Grass-type moves when the Pokémon's HP is low.","name":"Overgrow"},{"description":"Boosts the Pokémon's Speed stat in sunny weather.","hidden":true,"name":"Chlorophyll"}],"description":"Bulbasaur can be seen napping in bright sunlight. There is a seed on its back. By soaking up the sun's rays, the seed grows progressively larger.","evolution":{"evolves_to":{"id":2,"name":"Ivysaur"},"level":16},"id":1,"moves":[{"accuracy":100,"name":"Tackle","power":40,"pp":35,"type":"Normal"},{"accuracy":100,"name":"Vine Whip","power":45,"pp":25,"type":"Grass"},{"accuracy":90,"name":"Leech Seed","power":null,"pp":10,"type":"Grass"},{"accuracy":100,"name":"Growl","power":null,"pp":40,"type":"Normal"}],"name":"Bulbasaur","stats":{"attack":49,"defense":49,"hp":45,"special_attack":65,"special_defense":65,"speed":45},"type":["Grass","Poison"]}
---

[TestRunQuery/unpack_array - 1]
{
  "name": "bulbasaur",
  "url": "https://pokeapi.co/api/v2/pokemon/1/"
}
{
  "name": "ivysaur",
  "url": "https://pokeapi.co/api/v2/pokemon/2/"
}
{
  "name": "venusaur",
  "url": "https://pokeapi.co/api/v2/pokemon/3/"
}
{
  "name": "charmander",
  "url": "https://pokeapi.co/api/v2/pokemon/4/"
}
{
  "name": "charmeleon",
  "url": "https://pokeapi.co/api/v2/pokemon/5/"
}
{
  "name": "charizard",
  "url": "https://pokeapi.co/api/v2/pokemon/6/"
}
{
  "name": "squirtle",
  "url": "https://pokeapi.co/api/v2/pokemon/7/"
}
{
  "name": "wartortle",
  "url": "https://pokeapi.co/api/v2/pokemon/8/"
}
{
  "name": "blastoise",
  "url": "https://pokeapi.co/api/v2/pokemon/9/"
}
---

[TestRunQuery/compact_unpack_array - 1]
{"name":"bulbasaur","url":"https://pokeapi.co/api/v2/pokemon/1/"}
{"name":"ivysaur","url":"https://pokeapi.co/api/v2/pokemon/2/"}
{"name":"venusaur","url":"https://pokeapi.co/api/v2/pokemon/3/"}
{"name":"charmander","url":"https://pokeapi.co/api/v2/pokemon/4/"}
{"name":"charmeleon","url":"https://pokeapi.co/api/v2/pokemon/5/"}
{"name":"charizard","url":"https://pokeapi.co/api/v2/pokemon/6/"}
{"name":"squirtle","url":"https://pokeapi.co/api/v2/pokemon/7/"}
{"name":"wartortle","url":"https://pokeapi.co/api/v2/pokemon/8/"}
{"name":"blastoise","url":"https://pokeapi.co/api/v2/pokemon/9/"}
---

[TestRunQuery/it's_fucking_raw - 1]
bulbasaur
ivysaur
venusaur
charmander
charmeleon
charizard
squirtle
wartortle
blastoise
---

[TestRunQuery/cooked - 1]
"bulbasaur"
"ivysaur"
"venusaur"
"charmander"
"charmeleon"
"charizard"
"squirtle"
"wartortle"
"blastoise"
---

[TestRunQuery/slurp - 1]
[
  {
    "abilities": [
      {
        "description": "Powers up Grass-type moves when the Pokémon's HP is low.",
        "name": "Overgrow"
      },
      {
        "description": "Boosts the Pokémon's Speed stat in sunny weather.",
        "hidden": true,
        "name": "Chlorophyll"
      }
    ],
    "description": "Bulbasaur can be seen napping in bright sunlight. There is a seed on its back. By soaking up the sun's rays, the seed grows progressively larger.",
    "evolution": {
      "evolves_to": {
        "id": 2,
        "name": "Ivysaur"
      },
      "level": 16
    },
    "id": 1,
    "moves": [
      {
        "accuracy": 100,
        "name": "Tackle",
        "power": 40,
        "pp": 35,
        "type": "Normal"
      },
      {
        "accuracy": 100,
        "name": "Vine Whip",
        "power": 45,
        "pp": 25,
        "type": "Grass"
      },
      {
        "accuracy": 90,
        "name": "Leech Seed",
        "power": null,
        "pp": 10,
        "type": "Grass"
      },
      {
        "accuracy": 100,
        "name": "Growl",
        "power": null,
        "pp": 40,
        "type": "Normal"
      }
    ],
    "name": "Bulbasaur",
    "stats": {
      "attack": 49,
      "defense": 49,
      "hp": 45,
      "special_attack": 65,
      "special_defense": 65,
      "speed": 45
    },
    "type": [
      "Grass",
      "Poison"
    ]
  },
  {
    "abilities": [
      {
        "description": "Powers up Fire-type moves when the Pokémon's HP is low.",
        "name": "Blaze"
      },
      {
        "description": "Boosts the Pokémon's Special Attack stat in sunny weather, but it also loses HP every turn.",
        "hidden": true,
        "name": "Solar Power"
      }
    ],
    "description": "The flame on Charmander's tail indicates its life force. It burns more brightly if the Pokémon is healthy.",
    "evolution": {
      "evolves_to": {
        "id": 5,
        "name": "Charmeleon"
      },
      "level": 16
    },
    "id": 4,
    "moves": [
      {
        "accuracy": 100,
        "name": "Scratch",
        "power": 40,
        "pp": 35,
        "type": "Normal"
      },
      {
        "accuracy": 100,
        "name": "Ember",
        "power": 40,
        "pp": 25,
        "type": "Fire"
      },
      {
        "accuracy": 100,
        "name": "Growl",
        "power": null,
        "pp": 40,
        "type": "Normal"
      },
      {
        "accuracy": 100,
        "name": "Smokescreen",
        "power": null,
        "pp": 20,
        "type": "Normal"
      }
    ],
    "name": "Charmander",
    "stats": {
      "attack": 52,
      "defense": 43,
      "hp": 39,
      "special_attack": 60,
      "special_defense": 50,
      "speed": 65
    },
    "type": [
      "Fire"
    ]
  },
  {
    "abilities": [
      {
        "description": "Powers up Water-type moves when the Pokémon's HP is low.",
        "name": "Torrent"
      },
      {
        "description": "The Pokémon gradually regains HP in rain.",
        "hidden": true,
        "name": "Rain Dish"
      }
    ],
    "description": "Squirtle's shell is not only for protection. Its round shape reduces water resistance, allowing it to swim at high speeds.",
    "evolution": {
      "evolves_to": {
        "id": 8,
        "name": "Wartortle"
      },
      "level": 16
    },
    "id": 7,
    "moves": [
      {
        "accuracy": 100,
        "name": "Tackle",
        "power": 40,
        "pp": 35,
        "type": "Normal"
      },
      {
        "accuracy": 100,
        "name": "Water Gun",
        "power": 40,
        "pp": 25,
        "type": "Water"
      },
      {
        "accuracy": null,
        "name": "Withdraw",
        "power": null,
        "pp": 40,
        "type": "Water"
      },
      {
        "accuracy": 100,
        "name": "Bubble",
        "power": 40,
        "pp": 30,
        "type": "Water"
      }
    ],
    "name": "Squirtle",
    "stats": {
      "attack": 48,
      "defense": 65,
      "hp": 44,
      "special_attack": 50,
      "special_defense": 64,
      "speed": 43
    },
    "type": [
      "Water"
    ]
  }
]
---

[TestRunQuery/compact_and_slurp - 1]
[{"abilities":[{"description":"Powers up Grass-type moves when the Pokémon's HP is low.","name":"Overgrow"},{"description":"Boosts the Pokémon's Speed stat in sunny weather.","hidden":true,"name":"Chlorophyll"}],"description":"Bulbasaur can be seen napping in bright sunlight. There is a seed on its back. By soaking up the sun's rays, the seed grows progressively larger.","evolution":{"evolves_to":{"id":2,"name":"Ivysaur"},"level":16},"id":1,"moves":[{"accuracy":100,"name":"Tackle","power":40,"pp":35,"type":"Normal"},{"accuracy":100,"name":"Vine Whip","power":45,"pp":25,"type":"Grass"},{"accuracy":90,"name":"Leech Seed","power":null,"pp":10,"type":"Grass"},{"accuracy":100,"name":"Growl","power":null,"pp":40,"type":"Normal"}],"name":"Bulbasaur","stats":{"attack":49,"defense":49,"hp":45,"special_attack":65,"special_defense":65,"speed":45},"type":["Grass","Poison"]},{"abilities":[{"description":"Powers up Fire-type moves when the Pokémon's HP is low.","name":"Blaze"},{"description":"Boosts the Pokémon's Special Attack stat in sunny weather, but it also loses HP every turn.","hidden":true,"name":"Solar Power"}],"description":"The flame on Charmander's tail indicates its life force. It burns more brightly if the Pokémon is healthy.","evolution":{"evolves_to":{"id":5,"name":"Charmeleon"},"level":16},"id":4,"moves":[{"accuracy":100,"name":"Scratch","power":40,"pp":35,"type":"Normal"},{"accuracy":100,"name":"Ember","power":40,"pp":25,"type":"Fire"},{"accuracy":100,"name":"Growl","power":null,"pp":40,"type":"Normal"},{"accuracy":100,"name":"Smokescreen","power":null,"pp":20,"type":"Normal"}],"name":"Charmander","stats":{"attack":52,"defense":43,"hp":39,"special_attack":60,"special_defense":50,"speed":65},"type":["Fire"]},{"abilities":[{"description":"Powers up Water-type moves when the Pokémon's HP is low.","name":"Torrent"},{"description":"The Pokémon gradually regains HP in rain.","hidden":true,"name":"Rain Dish"}],"description":"Squirtle's shell is not only for protection. Its round shape reduces water resistance, allowing it to swim at high speeds.","evolution":{"evolves_to":{"id":8,"name":"Wartortle"},"level":16},"id":7,"moves":[{"accuracy":100,"name":"Tackle","power":40,"pp":35,"type":"Normal"},{"accuracy":100,"name":"Water Gun","power":40,"pp":25,"type":"Water"},{"accuracy":null,"name":"Withdraw","power":null,"pp":40,"type":"Water"},{"accuracy":100,"name":"Bubble","power":40,"pp":30,"type":"Water"}],"name":"Squirtle","stats":{"attack":48,"defense":65,"hp":44,"special_attack":50,"special_defense":64,"speed":43},"type":["Water"]}]
---
